api/v3/auth.go

```diff
@@ -0,0 +1,612 @@
+	token, err := a.signChallengeToken(u.UserName, u.EmailVerificationToken)
+	if err != nil {
+		l.Errorw("failed to generate email verification jwt", "error", err)
+		return nil, grpc.Errorf(codes.Internal, "failed to generate email verification jwt")
+	}
+	// format the url the user clicks to activate email
+	url := fmt.Sprintf("https://api.temporal.cloud/v3/auth/verify?user=%s&challenge%s", u.UserName, token)
+	if a.dev {
+		url = fmt.Sprintf("https://dev.api.temporal.cloud/v3/auth/verify?user=%s&challenge=%s", u.UserName, token)
+	}
+	// send email message to queue for processing
+	if err = a.emails.PublishMessage(queue.EmailSend{
+		Subject: "Temporal Email Verification",
+		Content: fmt.Sprintf("please click this %s to activate temporal email functionality",
+			fmt.Sprintf("<a href=\"%s\">link</a>", url)),
+		ContentType: "text/html",
+		UserNames:   []string{u.UserName},
+		Emails:      []string{u.EmailAddress},
+	}); err != nil {
+		l.Errorw(eh.QueuePublishError, "error", err)
+		return nil, grpc.Errorf(codes.Internal, "failed to send verification email")
+	}
+	l.Info("user account registered")
+
+	// generate default usage data
+	usage, err := a.usage.NewUsageEntry(u.UserName, models.Free)
+	if err != nil {
+		l.Errorw("unexpected error when generating user usage data",
+			"error", err)
+		return nil, grpc.Errorf(codes.Internal, "failed to generate usage limits for user")
+	}
+
+	// return relevant user data
+	return toUser(u, usage), nil
+}
+
+// Recover facilitates account recovery
+func (a *AuthService) Recover(ctx context.Context, req *auth.RecoverReq) (*auth.Empty, error) {
+	if req.GetEmailAddress() == "" {
+		return nil, grpc.Errorf(codes.InvalidArgument, "email cannot be empty")
+	}
+	user, err := a.users.FindByEmail(req.GetEmailAddress())
+	if err != nil {
+		return nil, grpc.Errorf(codes.NotFound, eh.UserSearchError)
+	}
+	if !user.EmailEnabled {
+		return nil, grpc.Errorf(codes.FailedPrecondition,
+			"account does not have email enabled - unfortunately for security reasons we can't assist in recovery")
+	}
+	var l = a.l.With("user", user.UserName)
+
+	switch req.GetType() {
+	case auth.RecoverReq_PASSWORD:
+		newPW, err := a.users.ResetPassword(user.UserName)
+		if err != nil {
+			l.Errorw("failed to reset password",
+				"error", err)
+			return nil, grpc.Errorf(codes.Internal, eh.PasswordResetError)
+		}
+		if err := a.emails.PublishMessage(queue.EmailSend{
+			Subject:     "TEMPORAL Password Reset",
+			Content:     fmt.Sprintf("your password is %s", newPW),
+			ContentType: "text/html",
+			UserNames:   []string{user.UserName},
+			Emails:      []string{user.EmailAddress},
+		}); err != nil {
+			l.Errorw("failed to send email to queue",
+				"error", err)
+			return nil, grpc.Errorf(codes.Internal, eh.QueuePublishError)
+		}
+		return &auth.Empty{}, nil
+
+	case auth.RecoverReq_USERNAME:
+		if err := a.emails.PublishMessage(queue.EmailSend{
+			Subject:     "TEMPORAL User Name Reminder",
```

Comment:
I think we should probably rename these instances of `TEMPORAL` to `Temporal`

Reply:
a2be798

Reply:
:fire: